client/fingerprint/vault.go

```diff
@@ -22,71 +23,90 @@ const (
 		}
-		f.client, err = vapi.NewClient(vaultConfig)
+		client, err = vapi.NewClient(vaultConfig)
 		if err != nil {
-			return fmt.Errorf("Failed to initialize Vault client: %s", err)
+			return fmt.Errorf("Failed to initialize Vault client for %s: %s", cfg.Name, err)
 		}
-		useragent.SetHeaders(f.client)
+		f.clients[cfg.Name] = client
+		useragent.SetHeaders(client)
 	}
 
 	// Connect to vault and parse its information
-	status, err := f.client.Sys().SealStatus()
+	status, err := client.Sys().SealStatus()
 	if err != nil {
-
 		// Print a message indicating that Vault is not available anymore
-		if f.lastState == vaultAvailable {
-			f.logger.Info("Vault is unavailable")
+		if lastState, ok := f.lastStates[cfg.Name]; !ok || lastState == vaultAvailable {
+			f.logger.Info("Vault is unavailable", "cluster", cfg.Name)
 		}
-		f.lastState = vaultUnavailable
+		f.lastStates[cfg.Name] = vaultUnavailable
 		return nil
 	}
 
-	resp.AddAttribute("vault.accessible", strconv.FormatBool(true))
-	// We strip the Vault prefix because < 0.6.2 the version looks like:
-	// status.Version = "Vault v0.6.1"
-	resp.AddAttribute("vault.version", strings.TrimPrefix(status.Version, "Vault "))
-	resp.AddAttribute("vault.cluster_id", status.ClusterID)
-	resp.AddAttribute("vault.cluster_name", status.ClusterName)
+	if cfg.Name == "default" {
+		resp.AddAttribute("vault.accessible", strconv.FormatBool(true))
+		resp.AddAttribute("vault.version", strings.TrimPrefix(status.Version, "Vault "))
+		resp.AddAttribute("vault.cluster_id", status.ClusterID)
+		resp.AddAttribute("vault.cluster_name", status.ClusterName)
+	} else {
+		resp.AddAttribute(fmt.Sprintf("vault.%s.accessible", cfg.Name), strconv.FormatBool(true))
+		resp.AddAttribute(fmt.Sprintf("vault.%s.version", cfg.Name), strings.TrimPrefix(status.Version, "Vault "))
+		resp.AddAttribute(fmt.Sprintf("vault.%s.cluster_id", cfg.Name), status.ClusterID)
+		resp.AddAttribute(fmt.Sprintf("vault.%s.cluster_name", cfg.Name), status.ClusterName)
+	}
 
 	// If Vault was previously unavailable print a message to indicate the Agent
 	// is available now
-	if f.lastState == vaultUnavailable {
-		f.logger.Info("Vault is available")
+	if lastState, ok := f.lastStates[cfg.Name]; !ok || lastState == vaultUnavailable {
+		f.logger.Info("Vault is available", "cluster", cfg.Name)
 	}
-	f.lastState = vaultAvailable
+	f.lastStates[cfg.Name] = vaultAvailable
 	resp.Detected = true
 	return nil
 }
 
 func (f *VaultFingerprint) Periodic() (bool, time.Duration) {
-	if f.lastState == vaultAvailable {
-		// Fingerprint infrequently once Vault is initially discovered with wide
-		// jitter to avoid thundering herds of fingerprints against central Vault
-		// servers.
-		return true, (30 * time.Second) + helper.RandomStagger(90*time.Second)
+	for _, lastState := range f.lastStates {
+		if lastState != vaultAvailable {
+			return true, 15 * time.Second
+		}
```

Comment:
Probably not worth changing anything, but I think this means that a single cluster not being available will cause all cluster to be fingerprinted at this accelerated interval.

Reply:
Yeah, the architecture of periodic fingerprinting means we'd end up spinning up a separate fingerprinter goroutine for each Vault (and Consul, once that's done). Maybe we could do something a little more sophisticated here though and fire the fingerprinter goroutine without necessarily sending the API calls to healthy Vaults on each pass.

Reply:
I gave this a go in 05a9047. The results are a little dissatisfying. The healthy Vaults effectively get checked on a fixed 45s interval when one of the other Vaults is unhealthy. But this avoids waking up the outer loop unnecessarily and avoids desync where healthy vaults get accidentally skipped for one cycle when the group as a whole skips from unhealthy to healthy.